fineract-provider/src/main/java/org/apache/fineract/infrastructure/crypt/service/EncryptionKeyStoreServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.infrastructure.crypt.service;
+
+import com.sun.jersey.spi.resource.Singleton;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.fineract.infrastructure.configuration.domain.ConfigurationDomainService;
+import org.apache.fineract.infrastructure.core.service.DateUtils;
+import org.apache.fineract.infrastructure.core.service.ThreadLocalContextUtil;
+import org.apache.fineract.infrastructure.crypt.domain.EncryptionKeys;
+import org.apache.fineract.infrastructure.crypt.utils.RSAEncryptionUtils;
+import org.joda.time.Seconds;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.cache.annotation.Caching;
+import org.springframework.stereotype.Service;
+
+/**
+ * @author manoj
+ */
+@Service
+@Singleton
+public class EncryptionKeyStoreServiceImpl  implements EncryptionKeyStoreService{
+    private Map<String, EncryptionKeys> keyMap =  new ConcurrentHashMap<>();
+
+
+    private final ConfigurationDomainService configurationDomainService;
+    private final RSAEncryptionUtils rsaEncryptionUtils;
+
+    @Autowired
+    public EncryptionKeyStoreServiceImpl(ConfigurationDomainService configurationDomainService, RSAEncryptionUtils rsaEncryptionUtils) {
+        this.configurationDomainService = configurationDomainService;
+        this.rsaEncryptionUtils = rsaEncryptionUtils;
+    }
+
+    @Caching(evict = {
+            @CacheEvict(value = "encryptionKeys", key = "T(org.apache.fineract.infrastructure.core.service.ThreadLocalContextUtil).getTenant().getTenantIdentifier().concat(#type)")})
+    public void storeKey(String type, EncryptionKeys key){
+        this.keyMap.put(getKeyType(type), key);
+    }
+
+    private EncryptionKeys retrieveValidKey(String type){
+
+        EncryptionKeys keys = getKeys(type);
+        if(keys == null){
+            return null;
+        }
+        //check validityOftheKey
+        //get validity time from config
+        Integer validUpto = configurationDomainService.retrieveEncKeyExpirySeconds(type);
+        if(validUpto.equals(-1)){
+            return keys;
+        } else {
+            Seconds seconds = Seconds.secondsBetween(keys.getCreatedDateTime(), DateUtils.getLocalDateTimeOfTenant());
+            if(seconds.getSeconds() < validUpto) {
+                return keys;
+            }
+
+        }
+        return null;
```

Comment:
as above... perhaps it's better to "handle" under whatever condition you may end up here?

Reply:
same as above